### PR TITLE
chore: Add tests for copy-before-allreduce and non-tensor input; fix duplicate init of communicators, destroy communicators in teardown

### PR DESCRIPTION
**Pull request by AndyUB**
## What is done?

A test is added for copying tensor before performing allreduce.
A test is added for throwing error when an input to allreduce is not a tensor.
Reverted: Fix: A communicator can be initialized multiple times before. A set is used to ensure exactly 1 initialization for each communicator.
Reverted: Fix: Collective communicators are not destroyed during teardown before. Now they are.
Reverted: Corresponding tests are added.
Chore: Tests are polished. Reduced the number of tests that require GPUs.

## What is experimental?

How do we check whether or not a communicator is already initialized when the user passes it in? One option is to explicitly ask the user to provide uninitialized communicators. Another option is to allow user to specify whether or not the communicator has been initialized.

Should we always destroy custom communicators during teardown? One option is to allow users to specify whether or not they want communicators to be destroyed during teardown.

(Stephanie) RE: Add a function that tells whether a `GPUCommunicator` is already initialized. Whoever initializes the `GPUCommunicator` destroys it.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py [hunk 1]
```diff
@@ -80,6 +80,10 @@ def compute_with_tuple_args(self, args, i: int):
         tensor = torch.ones(shape, dtype=dtype, device=self.device) * value
         return tensor
 
+    def recv_as_copy(self, tensor):
```
Comment (dengwxn) [inline python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: Do we need a separate recv func?

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py [hunk 2]
```diff
@@ -100,6 +104,72 @@ def forward(self, inp):
         return torch.randn(10, 10)
 
 
+class TestNcclGroup(GPUCommunicator):
```
Comment (dengwxn) [inline python/ray/dag/tests/experimental/test_torch_tensor_dag.py:110]: We can move this back to each test for now, because I can't pass tests when it's outside. I suspect there might be some environment problems.

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py (hunk 1 above)
Comment (AndyUB) [reply python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: Preferably yes, since the other `recv...()` fucntions don't return the tensor directly.

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py (hunk 2 above)
Comment (AndyUB) [reply python/ray/dag/tests/experimental/test_torch_tensor_dag.py:110]: ok

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py (hunk 1 above)
Comment (dengwxn) [reply python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: Rename to `recv_tensor` then.

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py [hunk 3]
```diff
@@ -1792,6 +1796,84 @@ def test_torch_tensor_nccl_all_reduce_scheduling_one_ready_group(ray_start_regul
...
+    assert torch.equal(result[2], expected_original_tensor_value)
+    assert torch.equal(result[3], expected_original_tensor_value)
+
+    compiled_dag.teardown()
+
+
+@pytest.mark.parametrize("ray_start_regular", [{"num_cpus": 4}], indirect=True)
+def test_torch_tensor_nccl_all_reduce_non_tensor_input(ray_start_regular):
+    """
+    Test that an error is thrown when an all-reduce takes non-tensor input.
```
Comment (dengwxn) [inline python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: Nit: inputs.

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py [hunk 4]
```diff
@@ -1792,6 +1796,84 @@ def test_torch_tensor_nccl_all_reduce_scheduling_one_ready_group(ray_start_regul
...
+
+    assert (
+        sum(node["Resources"].get("GPU", 0) for node in ray.nodes()) > 0
+    ), "This test requires at least 1 GPUs"
+
+    actor_cls = TorchTensorWorker.options(num_cpus=0, num_gpus=1)
+    worker = actor_cls.remote()
+
+    with InputNode() as inp:
+        nontensor = worker.send.bind((10,), torch.float16, inp, False)
```
Comment (dengwxn) [inline python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: Nit: non_tensor.
Why is this a non-tensor? What does `send` return?

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py [hunk 5]
```diff
@@ -1792,6 +1796,84 @@ def test_torch_tensor_nccl_all_reduce_scheduling_one_ready_group(ray_start_regul
...
+
+    num_workers = 2
+    workers = [actor_cls.remote() for _ in range(num_workers)]
+
+    shape = (10,)
+    dtype = torch.float16
+    with InputNode() as inp:
+        x = workers[0].send.bind(shape, dtype, inp)
+        y = workers[1].send.bind(shape, dtype, inp)
+
```
Comment (dengwxn) [inline python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: Nit: these three empty lines are not needed.

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py [hunk 6]
```diff
@@ -1792,6 +1796,84 @@ def test_torch_tensor_nccl_all_reduce_scheduling_one_ready_group(ray_start_regul
     compiled_dag.teardown()
 
 
+@pytest.mark.parametrize("ray_start_regular", [{"num_cpus": 4}], indirect=True)
+def test_torch_tensor_copy_before_nccl_all_reduce(ray_start_regular):
+    """
+    Test that a tensor is copied before all reduce is launched.
```
Comment (dengwxn) [inline python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: Nit: "before all-reduce is called`.

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py [hunk 7]
```diff
@@ -1792,6 +1796,84 @@ def test_torch_tensor_nccl_all_reduce_scheduling_one_ready_group(ray_start_regul
...
+    That is, the tensor is not all-reduced in place and can be reused
+    in the DAG as input to other nodes.
```
Comment (dengwxn) [inline python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: Nit: "as inputs to other nodes in the DAG."

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py (hunk 3 above)
Comment (AndyUB) [reply python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: ack

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py (hunk 4 above)
Comment (AndyUB) [reply python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: Changed to `worker.send.bind((10,), torch.float16, inp, send_tensor=False`.
`send` returns 1 if `send_tensor` is `False`.

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py (hunk 5 above)
Comment (AndyUB) [reply python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: ack

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py (hunk 6 above)
Comment (AndyUB) [reply python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: ack

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py (hunk 7 above)
Comment (AndyUB) [reply python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: ack

File: python/ray/dag/compiled_dag_node.py [hunk 8]
```diff
@@ -689,6 +689,9 @@ def __init__(
         # Uniquely identifies the NCCL communicator that will be used within
         # this DAG, if any.
         self._nccl_group_id: Optional[str] = None
+        # Identifies the set of NCCL communicators used within this DAG,
+        # including both collective and P2P communicators.
```
Comment (dengwxn) [inline python/ray/dag/compiled_dag_node.py]: Nit: both P2P and collective.

File: python/ray/experimental/channel/gpu_communicator.py [hunk 9]
```diff
@@ -130,3 +130,6 @@ def destroy() -> None:
         done here. Implement as a noop is nothing is needed.
         """
         raise NotImplementedError
+
+    def __deepcopy__(self, _) -> "GPUCommunicator":
```
Comment (dengwxn) [inline python/ray/experimental/channel/gpu_communicator.py]: Seems weird. It does not actually copy. What's the use case for this?

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py [hunk 10]
```diff
@@ -1119,41 +1127,23 @@ def test_torch_tensor_nccl_all_reduce_custom_comm_wrong_actors(ray_start_regular
...
-    ), "This test requires at least 2 GPUs"
-
-    actor_cls = TorchTensorWorker.options(num_cpus=0, num_gpus=1)
+    actor_cls = TorchTensorWorker.options(num_gpus=0)
 
     num_workers = 2
     workers = [actor_cls.remote() for _ in range(num_workers)]
 
-    class TestNcclGroup(GPUCommunicator):
+    class DummyNcclGroup(GPUCommunicator):
```
Comment (dengwxn) [inline python/ray/dag/tests/experimental/test_torch_tensor_dag.py:1074]: Use `MockNcclGroup`.

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py [hunk 11]
```diff
@@ -1218,14 +1201,7 @@ def test_torch_tensor_nccl_all_reduce_duplicate_actors(ray_start_regular):
...
     """
-    if not USE_GPU:
-        pytest.skip("NCCL tests require GPUs")
-
-    assert (
-        sum(node["Resources"].get("GPU", 0) for node in ray.nodes()) > 1
-    ), "This test requires at least 2 GPUs"
-
-    actor_cls = TorchTensorWorker.options(num_cpus=0, num_gpus=1)
+    actor_cls = TorchTensorWorker.options(num_gpus=0)
```
Comment (dengwxn) [inline python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: Does it work without `num_gpus=0`?

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py [hunk 12]
```diff
@@ -986,7 +988,9 @@ def test_torch_tensor_nccl_all_reduce_get_partial(ray_start_regular):
             for i, worker in enumerate(workers)
         ]
         collectives = collective.allreduce.bind(computes, ReduceOp.SUM)
-        dag = workers[0].recv.bind(collectives[0])
+        recv = workers[0].recv.bind(collectives[0])
+        tensor = workers[1].recv_tensor.bind(collectives[0])
```
Comment (dengwxn) [inline python/ray/dag/tests/experimental/test_torch_tensor_dag.py:996]: Is the change of adding `tensor` is testing both metadata and tensor, while previously only testing metadata?

File: python/ray/dag/compiled_dag_node.py (hunk 8 above)
Comment (AndyUB) [reply python/ray/dag/compiled_dag_node.py]: done

File: python/ray/experimental/channel/gpu_communicator.py (hunk 9 above)
Comment (AndyUB) [reply python/ray/experimental/channel/gpu_communicator.py]: dag_node.py:104: `self._type_hint = copy.deepcopy(typ)` copies an entire type hint. Since a `GPUCommunicator` is contained in some type hint, it is also copied.
If the same `GPUCommunicator` is specified for multiple type hints, it will be copied multiple times and the different copies won't be equal to each other. That means for each copy of the same `GPUCommunicator`, a new NCCL group will be initialized during DAG compilation. This does not deduplicate the `GPUCommunicator`s.
The hack here does not actually copy so that if the same `GPUCommunicator` is specified for multiple type hints, after copying the type hints, their `GPUCommunicator`s are still the same one.
Another approach would be to change `__eq__`, but `GPUCommunicator` is an interface. If we require the user to provide an `__eq__` for each class that inherits from `GPUCommunicator`, it feels a bit complicated and may be error-prone.
Maybe there are other ways to do this?

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py (hunk 10 above)
Comment (AndyUB) [reply python/ray/dag/tests/experimental/test_torch_tensor_dag.py:1074]: This change is a response to Stephanie's comment about reducing the number of tests that require GPUs. Because `MockNcclGroup` uses GPU, switching to `MockNcclGroup` requires this test to use GPUs.

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py (hunk 11 above)
Comment (AndyUB) [reply python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: Yes, it works without `num_gpus=0`. Stephanie commented that this test is only testing for errors and should not require GPU, so I changed it.

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py (hunk 12 above)
Comment (AndyUB) [reply python/ray/dag/tests/experimental/test_torch_tensor_dag.py:996]: Yes. Stephanie commented that we should check both.

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py (hunk 11 above)
Comment (AndyUB) [reply python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: Remove `num_gpus=0`.

File: python/ray/experimental/channel/gpu_communicator.py (hunk 9 above)
Comment (AndyUB) [reply python/ray/experimental/channel/gpu_communicator.py]: Currently reverted.
